hbase-server/src/main/java/org/apache/hadoop/hbase/replication/regionserver/ReplicationSink.java

```diff
@@ -265,6 +266,11 @@ public void replicateEntries(List<WALEntry> entries, final CellScanner cells,
               mutation.setClusterIds(clusterIds);
               mutation.setAttribute(ReplicationUtils.REPLICATION_ATTR_NAME,
                 HConstants.EMPTY_BYTE_ARRAY);
+              if (attributeList != null) {
```

Comment:
This is WAL extended attribute, not mutation attribute? I do not think we should just set them as attributes here...

Reply:
There has to be a consumer of WAL extended attribute. The primary purpose for anything associated with WALKey is meant for deriving appropriate Mutation behavior or replication related behavior (like list of cluster ids). Hence if we allow coproc to set the WAL extended attributes, we should also let the attributes be consumed at sink side. Otherwise we won't be able to have any consumer of these attributes, correct?

Reply:
The goal of the WAL extended attribute is to pass some attributes down to the replication implementation from the coproc implementation, not to add it to the mutation's attribute. We should not mix things up.

Reply:
> The goal of the WAL extended attribute is to pass some attributes down to the replication implementation from the coproc implementation

That is correct but coproc would add it only so that it can consume it at sink side correct? Otherwise who else would read WAL attributes from WALEntry at sink side? HBase does it, while preparing for sink side mutations.

Reply:
If required, we can make this behavior configurable?

Reply:
No, you do not understand the problem here. While you say that

> the sink side mutation attributes can be expected to be set up by HBase since hbase prepares mutations from WAL entries.

But will we set these attributes to mutation's attributes at source cluster? If not, why should we set them at the sink cluster?

Reply:
> > The goal of the WAL extended attribute is to pass some attributes down to the replication implementation from the coproc implementation
> 
> That is correct but coproc would add it only so that it can consume it at sink side correct? Otherwise who else would read WAL attributes from WALEntry at sink side? HBase does it, while preparing for sink side mutations.

The original design is to read it in replication endpoint, IIRC. And also, you can read it in ReplicationSink at the sink cluster, no problem.

The problem here is you want to add the WAL attributes to mutation's attributes. This is a no no. We should not mix things up.

Reply:
And if phoenix wants this feature, please describe more about at lease one usage. Let's see how to better support it.

Reply:
How do you read these attributes at the source cluster's coproc?

Reply:
> And if phoenix wants this feature, please describe more about at lease one usage. Let's see how to better support it.

So let me take a simple example. Phoenix supports multi-tenancy by providing multi-tenant connection. Hence, same table can be shared by multiple tenants. This requires tenant id to be a rowkey prefix in hbase rowkey. When a particular tenant creates tenant connection and writes data, those data are not visible to other tenants.
Tenant id is very basic entity in phoenix. Now let's say we want to introduce some level of caching in phoenix server side, the cache would contain tenant id.

For HBase sink cluster, mutation is already created for specific table because table name is available in WALKey. For phoenix, let's say table name is available and mutation has full rowkey as well but it's very difficult to derive tenant id from this rowkey. Hence, phoenix coproc needs to know of tenant id attribute with mutation so that it can form tenant level connection or derive further attributes at sink side and so on.

In addition to tenant id, there are a few more imp metadata that is required at sink side coproc, without which it's not possible to be aware of how the data is meant to be used.

> How do you read these attributes at the source cluster's coproc?

If we take same example as above, multi tenant connection is already created by client and hence tenant id is always available with all operations performed at source cluster coproc. Besides some of the imp attributes are already attached as mutation attributes at source side. But when the mutations are generated at sink side, while hbase metadata is available, coproc like phoenix metadata is not available and it would be great if we can attach these metadata attributes to sink mutation, otherwise sink coproc would not be able to derive them.
Making this configurable feature with proper documentation is also fine, if you don't mind. WDYT?

Reply:
For the sake of this example, let's say that tenant id (part of the extended attribute in WALKey) is as important to phoenix, as is rowkey to hbase. It's quite basic entity.